lib/iris/coords.py

```diff
@@ -52,6 +52,697 @@
+    """
+    Superclass for dimensional metadata.
+
+    """
+
+    _MODE_ADD = 1
+    _MODE_SUB = 2
+    _MODE_MUL = 3
+    _MODE_DIV = 4
+    _MODE_RDIV = 5
+    _MODE_SYMBOL = {_MODE_ADD: '+', _MODE_SUB: '-',
+                    _MODE_MUL: '*', _MODE_DIV: '/',
+                    _MODE_RDIV: '/'}
+
+    def __init__(self, values, standard_name=None, long_name=None,
+                 var_name=None, units='no-unit', attributes=None):
+        """
+        Constructs a single dimensional metadata object.
+
+        Args:
+
+        * values:
+            The values of the dimensional metadata.
+
+        Kwargs:
+
+        * standard_name:
+            CF standard name of the dimensional metadata.
+        * long_name:
+            Descriptive name of the dimensional metadata.
+        * var_name:
+            The netCDF variable name for the dimensional metadata.
+        * units
+            The :class:`~cf_units.Unit` of the dimensional metadata's values.
+            Can be a string, which will be converted to a Unit object.
+        * attributes
+            A dictionary containing other cf and user-defined attributes.
+
+        """
+        #: CF standard name of the quantity that the metadata represents.
+        self.standard_name = standard_name
+
+        #: Descriptive name of the metadata.
+        self.long_name = long_name
+
+        #: The netCDF variable name for the metadata.
+        self.var_name = var_name
+
+        #: Unit of the quantity that the metadata represents.
+        self.units = units
+
+        #: Other attributes, including user specified attributes that
+        #: have no meaning to Iris.
+        self.attributes = attributes
+
+        # Set up DataManager attributes and values.
+        self._values_dm = None
+        self._values = values
+
+    def __getitem__(self, keys):
+        """
+        Returns a new dimensional metadata whose values are obtained by
+        conventional array indexing.
+
+        .. note::
+
+            Indexing of a circular coordinate results in a non-circular
+            coordinate if the overall shape of the coordinate changes after
+            indexing.
+
+        """
+        # Fetch the values.
+        values = self._values_dm.core_data()
+
+        # Index values with the keys.
+        _, values = iris.util._slice_data_with_keys(
+            values, keys)
+
+        # Copy values after indexing to avoid making metadata that is a
+        # view on another metadata.
+        values = values.copy()
+
+        # If the metadata is a coordinate and it has bounds, repeat the above
+        # with the bounds.
+        if isinstance(self, Coord):
+            if self.has_bounds():
+                bounds = self._bounds_dm.core_data()
+                _, bounds = iris.util._slice_data_with_keys(bounds, keys)
+                bounds = bounds.copy()
+            else:
+                bounds = None
+
+            copy_args = dict(points=values, bounds=bounds)
```

Comment:
Just as this code section is the *only* place where you allow the _DimensionalMetadata code to refer to _bounds_dm,  so it is the only branch that sets up to pass a 'bounds' keyword to the copy function.
This reads oddly, as the local 'copy' definition, immediately below, does not support a bounds keyword.
So I think it is worth an extra comment here, to make it clear that 'self' here _must_ be a Coord, so that makes the difference.

Reply:
Additional... this is ***not*** the only place where _DimensionMetadata deals with bounds : it also happens in `convert_units`.
But all the other methods, like `__getitem__`, seem to use an "override, call parent + extend" strategy instead where bounds must be handled.
Not sure what I think of this yet.